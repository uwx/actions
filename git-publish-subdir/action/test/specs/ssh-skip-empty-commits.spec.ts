import { promises as fs } from 'fs';
import * as path from 'path';
import { mkdirP } from '@actions/io';

import * as util from '../util';
import { prepareTestFolders } from '../util/io';

it('Skip empty commits', async () => {
  const folders = await prepareTestFolders({ __filename });

  await util.wrappedExec('git init --bare', { cwd: folders.repoDir });

  // Create dummy data
  await mkdirP(path.join(folders.dataDir, 'dummy'));
  await fs.writeFile(path.join(folders.dataDir, 'dummy', 'baz'), 'foobar');
  await fs.writeFile(path.join(folders.dataDir, 'dummy', '.bat'), 'foobar');

  // Run Action
  await util.runWithGithubEnv(
    path.basename(__filename),
    {
      REPO: folders.repoUrl,
      BRANCH: 'branch-a',
      FOLDER: folders.dataDir,
      SSH_PRIVATE_KEY: (await fs.readFile(util.SSH_PRIVATE_KEY)).toString(),
      KNOWN_HOSTS_FILE: util.KNOWN_HOSTS,
      SKIP_EMPTY_COMMITS: 'true',
    },
    's0/test',
    {},
    's0'
  );
  const fullSha1 = await util.getFullRepoSha();
  // Change files and run action again
  await fs.writeFile(path.join(folders.dataDir, 'dummy', 'bat'), 'foobar');
  await util.runWithGithubEnv(
    path.basename(__filename),
    {
      REPO: folders.repoUrl,
      BRANCH: 'branch-a',
      FOLDER: folders.dataDir,
      SSH_PRIVATE_KEY: (await fs.readFile(util.SSH_PRIVATE_KEY)).toString(),
      KNOWN_HOSTS_FILE: util.KNOWN_HOSTS,
      SKIP_EMPTY_COMMITS: 'true',
    },
    's0/test',
    {},
    's0'
  );
  const fullSha2 = await util.getFullRepoSha();
  // Run the action again with no content changes to test skip behaviour
  await util.runWithGithubEnv(
    path.basename(__filename),
    {
      REPO: folders.repoUrl,
      BRANCH: 'branch-a',
      FOLDER: folders.dataDir,
      SSH_PRIVATE_KEY: (await fs.readFile(util.SSH_PRIVATE_KEY)).toString(),
      KNOWN_HOSTS_FILE: util.KNOWN_HOSTS,
      SKIP_EMPTY_COMMITS: 'true',
    },
    's0/test',
    {},
    's0'
  );

  // Check that the log of the repo is as expected
  // (check tree-hash, commit message, and author)
  // TODO: test {msg} placeholder and running action outside of a git repo
  let log = (
    await util.exec(
      'git log --pretty="format:msg:%B%ntree:%T%nauthor:%an <%ae>" branch-a',
      {
        cwd: folders.repoDir,
      }
    )
  ).stdout;
  const sha1 = fullSha1.substr(0, 7);
  const sha2 = fullSha2.substr(0, 7);
  const cleanedLog = log.replace(sha1, '<sha1>').replace(sha2, '<sha2>');
  expect(cleanedLog).toMatchSnapshot();
});
