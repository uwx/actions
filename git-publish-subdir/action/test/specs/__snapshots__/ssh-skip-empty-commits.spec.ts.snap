// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Skip empty commits 1`] = `
"msg:Update branch-a to output generated at <sha1>

tree:b60735c9b4d9728b47c0f2752bb973cd6f3d9d80
author:s0 <s0@users.noreply.github.com>
msg:Update branch-a to output generated at <sha2>

tree:8bf87c66655949e66937b11593cc4ae732d1f610
author:s0 <s0@users.noreply.github.com>"
`;
